Jenkinsfile

```diff
@@ -1,19 +1,14 @@
-def odsImageTag = env.ODS_IMAGE_TAG ?: 'latest'
-def odsGitRef = env.ODS_GIT_REF ?: 'production'
+def odsImageTag
 def final projectId = 'prov'
 def final componentId = 'prov-app'
 def final credentialsId = "${projectId}-cd-cd-user-with-password"
-def sharedLibraryRepository
 def dockerRegistry
 node {
-  sharedLibraryRepository = env.SHARED_LIBRARY_REPOSITORY
+  odsImageTag = env.ODS_IMAGE_TAG ?: 'latest'
   dockerRegistry = env.DOCKER_REGISTRY
 }
 
-library identifier: "ods-library@${odsGitRef}", retriever: modernSCM(
-  [$class: 'GitSCMSource',
-   remote: sharedLibraryRepository,
-   credentialsId: credentialsId])
+@Library('ods-jenkins-shared-library@2.x') _
```

Comment:
Why is this not using the value from `env.ODS_GIT_REF`? Is this what you just told me about the params issue? Or am I confused now? It would be great not to have specific versions here to avoid merge conflicts and having stale versions because we forget to update ...

Reply:
@michaelsauter  - this is what i mean -NOT working .. I tried all sorts of things .. and this does not work :(

Reply:
Ok got it. I would like to have some time this week to experiment a little. Maybe there is a way to get it to work.

Reply:
enjoy (I tried with params / properties / .... :))

Reply:
Okay I gave this a spin. It is, as you figured out, not possible with `@Library`. However, it is possible using `library`. See https://jenkins.io/doc/book/pipeline/shared-libraries/#loading-libraries-dynamically. I tried that and it worked.

From what I understand, the main difference is:

> Whereas the @Library annotation prepares the “classpath” of the script prior to compilation, by the time a library step is encountered the script has already been compiled. Therefore you cannot import or otherwise “statically” refer to types from the library.

Which simply means that one cannot use classes from the `src/` directory well.

I want to give this some more thought, but I think I am OK with that restriction:
- we expose the main building blocks under `vars/` anyway
- people can still change to the `@Library` annotation if needed, and then just hardcode the version they need

Thoughts?

Reply:
so - no real drawbacks then? ... what snippet did you make fly?

Reply:
I made up my mind. My recommendation is this:

* Use `@Library('ods-jenkins-shared-library@@ods_git_ref@') _` in the `Jenkinsfile.template` of quickstarters, which will render to e.g. `@Library('ods-jenkins-shared-library@2.x') _`
* Use `library("ods-jenkins-shared-library@${odsGitRef}")` in files like here where we want to dynamically resolve. This style would be used for the `Jenkinsfile` (the provisioning one) in the quickstarters

Reply:
@michaelsauter  - you rock! ... changed and pushed :)